### DISCUSSION TITLE
How to use I2C to read/write register with 10-bit register address?

### DISCUSSION THREAD

**Discussion by snowuyl** [zephyrproject-rtos/zephyr#62805] [Q&A]
**Describe the bug**
I try to read/write registers of TI LP5861 through I2C interface. But it failed with error code -5. Does Zephyr project provide sample code to read register with 10-bit register address ?
**Expected behavior**
Can reading/writing registers of TI LP5861 through I2C successfully.

**Logs and console output**
00> *** Booting Zephyr OS build v3.3.99-ncs1-1 ***
00> Starting button Test example
00> device_is_ready() ok
00> nrfx_gpiote_input_configure() okay for button
00> val=1
00> button released
00> start a timer
00> now-1-11-30 0:0:0
00> 0
00> ledDriverInit
00> tca6424aReadReg(regAddr=0x0)
00> LP5861_REG_ADDR_HIGH(regAddr)=0x0, LP5861_REG_ADDR_LOW(regAddr)=0x0
00> devAddr=0x81, startAddr=0x0
00> [00:00:00.501,129] <err> i2c_nrfx_twim: Error on I2C line occurred for message 0
00> i2c_burst_read(0x0) failed, ret=-5
00> _buttonTimerHandler: 3000
00> now-1-11-30 0:0:0
00> button released

**Environment (please complete the following information):**
 - OS: Windows 11
 - Toolchain: Zephyr OS build v3.3.99-ncs1-1

**Comment by snowuyl** [comment]
<img width="815" alt="I2c_data_format" src="https://github.com/zephyrproject-rtos/zephyr/assets/8336439/a66123e3-f027-45c3-a013-7b3881b39389">
<img width="815" alt="I2c_timing" src="https://github.com/zephyrproject-rtos/zephyr/assets/8336439/bc7078e2-ddc7-437e-a724-d226c4179974">
<img width="806" alt="registers" src="https://github.com/zephyrproject-rtos/zephyr/assets/8336439/da6ef085-0ae8-486f-88eb-34ae973e09c4">

**Comment by snowuyl** [comment]
[LP5861.pdf](https://github.com/zephyrproject-rtos/zephyr/files/12655418/LP5861.pdf)

**Comment by Prasanna1717** [comment]
Hello,
I wrote a code for reading and writing to a register using Zephyr's i2c API. It's my old sample code, please feel free to use it for a test case and let me know.
Please replace "I2C_1" with the name of your specific I2C controller and 0xABC with the actual 10-bit address of your LP5861 device.I guess, It should solve your bug.

CODE,
#include <zephyr.h>
#include <device.h>
#include <drivers/i2c.h>
#include <stdio.h>

#define I2C_DEV_NAME "I2C_1"  // Replace with your I2C controller's name
#define LP5861_10BIT_ADDR 0xABC  // Replace with your LP5861's 10-bit address

void main(void) {
    struct device *i2c_dev;
    uint8_t data_to_write = 0x34;  // Data to write to the register
    uint8_t read_data;             // Data read from the register
    uint8_t register_address = 0xAB; // Replace with the actual register address
    struct i2c_msg msgs[2];

    // Open the I2C device
    i2c_dev = device_get_binding(I2C_DEV_NAME);
    if (!i2c_dev) {
        printk("I2C device not found\n");
        return;
    }

    // Prepare the write message
    msgs[0].buf = &data_to_write;
    msgs[0].len = sizeof(data_to_write);
    msgs[0].flags = I2C_MSG_WRITE;

    // Prepare the read message
    msgs[1].buf = &read_data;
    msgs[1].len = sizeof(read_data);
    msgs[1].flags = I2C_MSG_READ | I2C_MSG_STOP;

    // Start the I2C communication
    int ret = i2c_transfer(i2c_dev, &msgs[0], 1, LP5861_10BIT_ADDR);
    if (ret) {
        printk("Error writing to register: %d\n", ret);
        return;
    }

    // Delay (if needed) to ensure the write operation has taken effect

    // Start the I2C communication for read
    ret = i2c_transfer(i2c_dev, &msgs[1], 1, LP5861_10BIT_ADDR);
    if (ret) {
        printk("Error reading from register: %d\n", ret);
        return;
    }

    // Print the read data
    printk("Read data from register: 0x%02X\n", read_data);
}

Regards
Prasanna

**Comment by ndrs-pst** [comment]
@snowuyl, I hope you've managed to solve your issue.

For anyone working on projects that require 10-bit addressing in I2C, here's some helpful information.

---
To effectively experiment with I2C drivers, make sure to enable `CONFIG_I2C_SHELL=y` and use i2c command in console.

When dealing with 10-bit register addresses, setting `I2C_MSG_ADDR_10_BITS` is crucial.
This is accomplished by calling `i2c_configure()`. The code snippet provided below illustrates the process:

```
int ret;
uint32_t dev_config = 0;

ret = i2c_get_config(dev, &dev_config);
dev_config |= I2C_MSG_ADDR_10_BITS;          /* Enabling the 10-bit address mode */
ret = i2c_configure(dev, dev_config);        /* Configuring the device with the new setting */
```

**Comment by stichkule** [comment]
@snowuyl Not sure if you're issue is resolved, but looking at that I2C spec for LP5861, it doesn't look like 10-bit addressing. For 10-bit addressing, the first byte MUST contain 0b11110XX (XX represents 2 bits from the slave address). Followed by the remaining 8 bits of the slave address in the second byte.

See https://www.totalphase.com/support/articles/200349176-7-bit-8-bit-and-10-bit-i2c-slave-addressing/#10bit
